rfc/20250711-Middleware-For-Enhanced-Operations.md

```diff
@@ -0,0 +1,378 @@
+
+1. **Project-level middleware executes first** - Applied to all resources regardless of provider
+2. **Provider-level middleware executes second** - Applied only to resources from that provider
+
+This hierarchy allows for:
+- Global policies to be enforced before provider-specific checks
+- Provider middleware to see and build upon project middleware metadata
+- Layered validation from general to specific
+
+#### Duplicate Middleware Handling
+
+If the same middleware is attached at both levels, it will execute **twice** for resources from that provider:
+
+```hcl
+middleware "cost_estimator" {
+  command = "cost-estimate"
+}
+
+terraform {
+  middleware = [middleware.cost_estimator]  # First execution
+}
+
+provider "aws" {
+  middleware = [middleware.cost_estimator]  # Second execution - overwrites metadata!
+}
+```
+
+In this case:
+- The middleware runs twice for AWS resources
+- The second execution's metadata **overwrites** the first
+- Each execution has no knowledge it's being run multiple times
+
+Users should generally avoid this pattern unless the middleware is designed to handle multiple executions.
+
+### Communication
+As mentioned above, I propose that to keep it as open as possible we should use JSON-RPC 2.0 over stdio for communication with the possibility to open up to other transports in the future.
+
+#### Initialize Handshake
+```json
+// Request
+{
+  "jsonrpc": "2.0",
+  "method": "initialize",
+  "params": {
+    "version": "1.0",
+    "name": "cost_estimator"
+  },
+  "id": 1
+}
+
+// Response
+{
+  "jsonrpc": "2.0",
+  "id": 1,
+  "result": {
+    "capabilities": ["pre-plan", "post-plan", "pre-apply", "post-apply"]
+  }
+}
+```
+
+#### Hook Execution
+```json
+// Post-Plan Hook Request
+{
+  "jsonrpc": "2.0",
+  "method": "post-plan",
+  "params": {
+    "provider": "registry.opentofu.org/hashicorp/aws",
+    "resource_type": "aws_instance",
+    "resource_name": "example",
+    "resource_mode": "managed",
+    "planned_action": "Create",
+    "before": null,
+    "after": { /* planned state */ },
+    "config": { /* resource config */ }
+    ....
+  },
+  "id": 2
+}
+
+// Hook Response
+{
+  "jsonrpc": "2.0",
+  "id": 2,
+  "result": {
+    "status": "success",
+    "message": "Estimated monthly cost: $73.58",
+    "metadata": {
+      "estimated_cost": {
+        "hourly": 0.102,
+        "monthly": 73.58,
+        "currency": "USD"
+      }
+    }
+  }
+}
+```
+
+### Process Lifecycle
+
+Middleware processes are managed with a simple lifecycle:
+
+1. **Startup**: When OpenTofu begins execution (init, plan, apply, etc.), all configured middleware processes are started
+2. **Initialization**: Each middleware receives an initialize request and must respond with its capabilities
+3. **Operation**: Middleware processes remain running for the entire duration of the OpenTofu command
+4. **Shutdown**: When OpenTofu completes (successfully or with errors), all middleware processes are terminated
```

Comment:
Should we consider when a middleware wants to run through all lifecycles and then wants to do some functionality during shutdown?

I know you mentioned multiple invocations above, it not being aware of previous invocations of a middleware. But Im imagining something like some security middleware wants to ensure the plan is the same between planning and applying. They could even allow the user to still apply the different plan but during shutdown they show some warning or something. Just thinking out loud.